virtualizarr/writers/icechunk.py

```diff
@@ -96,65 +112,147 @@ def write_variables_to_icechunk_group(
+            append_dim=append_dim,
         )
 
 
-def write_variable_to_icechunk(
-    store: "IcechunkStore",
+def num_chunks(
+    array,
+    axis: int,
+):
+    return array.shape[axis] // array.chunks[axis]
+
+
+def resize_array(
     group: "Group",
     name: str,
     var: Variable,
-) -> None:
-    """Write a single (possibly virtual) variable into an icechunk store"""
-    if isinstance(var.data, ManifestArray):
-        write_virtual_variable_to_icechunk(
-            store=store,
-            group=group,
-            name=name,
-            var=var,
-        )
-    else:
-        raise ValueError(
-            "Cannot write non-virtual variables as virtual variables to Icechunk stores"
-        )
+    append_axis: int,
+) -> Array:
+    existing_array = group[name]
+    # This is the second time we check if the array is an instance of zarr.core.Array
+    # but it's necessary to ensure .shape and .resize are available
+    if not isinstance(existing_array, Array):
+        raise ValueError("Expected existing array to be a zarr.core.Array")
+    new_shape = list(existing_array.shape)
+    new_shape[append_axis] += var.shape[append_axis]
+    existing_array.resize(tuple(new_shape))
+    return existing_array
+
+
+def get_axis(
+    dims: list[str],
+    dim_name: Optional[str],
+) -> int:
+    if dim_name is None:
+        raise ValueError("dim_name must be provided")
+    return dims.index(dim_name)
+
+
+def _check_compatible_arrays(
+    ma: ManifestArray, existing_array: zarr.core.array.Array, append_axis: int
+):
+    manifest_api.check_combinable_zarr_arrays([ma, existing_array])
+    manifest_api.check_same_ndims([ma.ndim, existing_array.ndim])
+    arr_shapes = [ma.shape, existing_array.shape]
+    manifest_api.check_same_shapes_except_on_concat_axis(arr_shapes, append_axis)
+
+
+def _check_compatible_encodings(encoding1, encoding2):
+    for key, value in encoding1.items():
+        if key in encoding2:
+            if encoding2[key] != value:
+                raise ValueError(
+                    f"Cannot concatenate arrays with different values for encoding key {key}: {encoding2[key]} != {value}"
+                )
 
 
 def write_virtual_variable_to_icechunk(
     store: "IcechunkStore",
     group: "Group",
     name: str,
     var: Variable,
+    append_dim: Optional[str] = None,
 ) -> None:
     """Write a single virtual variable into an icechunk store"""
     ma = cast(ManifestArray, var.data)
     zarray = ma.zarray
+    mode = store.mode.str
 
-    # creates array if it doesn't already exist
-    arr = group.require_array(
-        name=name,
-        shape=zarray.shape,
-        chunk_shape=zarray.chunks,
-        dtype=encode_dtype(zarray.dtype),
-        codecs=zarray._v3_codec_pipeline(),
-        dimension_names=var.dims,
-        fill_value=zarray.fill_value,
-        # TODO fill_value?
-    )
+    dims: list[str] = cast(list[str], list(var.dims))
+    append_axis, existing_num_chunks, arr = None, None, None
```

Comment:
This line seems very weird - `append_axis` could be set to `None` within the `else` below, `existing_num_chunks` could maybe just be 0? and I don't see why you need to set `arr` here at all